src/utxo/db.rs

```diff
@@ -0,0 +1,132 @@
+use crate::bitcoin::consensus::serialize;
+use crate::bitcoin::{Block, OutPoint, TxOut};
+use crate::utxo::{Hash32, Hash64, Utxo};
+use bitcoin::consensus::deserialize;
+use log::debug;
+use rocksdb::{DBCompressionType, Options, WriteBatch, DB};
+use std::collections::HashMap;
+use std::convert::TryInto;
+use std::path::Path;
+
+pub struct DbUtxo {
+    db: DB,
+    updated_up_to_height: i32,
+    inserted_outputs: u64,
+}
+
+impl DbUtxo {
+    pub fn new<P: AsRef<Path>>(path: P) -> Result<DbUtxo, rocksdb::Error> {
+        let mut options = Options::default();
+        options.set_compression_type(DBCompressionType::Snappy);
+        options.create_if_missing(true);
+        let db = DB::open(&options, path)?;
+
+        let updated_up_to_height = db
+            .get("updated_up_to_height")?
+            .map(|e| e.try_into().unwrap())
+            .map(|e| i32::from_ne_bytes(e))
+            .unwrap_or(-1);
+
+        Ok(DbUtxo {
+            db,
+            updated_up_to_height,
+            inserted_outputs: 0,
+        })
+    }
+}
+
+impl Utxo for DbUtxo {
+    fn add(&mut self, block: &Block, height: u32) {
+        let height = height as i32;
+        debug!(
+            "height: {} updated_up_to: {}",
+            height, self.updated_up_to_height
+        );
+        if height > self.updated_up_to_height {
+            // since we can spend outputs created in this same block, we first put outputs in memory...
+            let total_outputs = block.txdata.iter().map(|e| e.output.len()).sum();
+            let mut current_outputs = HashMap::with_capacity(total_outputs);
```

Comment:
`current_outputs` -> `block_unspent_outputs` maybe? The name didn't give me much until I read through the whole code.

Reply:
What do you think about just `block_outputs`?  They may be spent in the same block...